src/v/iceberg/time_transform_visitor.cc

```diff
@@ -0,0 +1,38 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/time_transform_visitor.h"
+
+#include "iceberg/values.h"
+
+namespace iceberg {
+
+namespace {
+int32_t micros_to_hr(int64_t micros) {
+    static constexpr int64_t s_per_hr = 3600;
+    static constexpr int64_t micros_per_s = 1000000;
+    static constexpr int64_t micros_per_hr = micros_per_s * s_per_hr;
+    return static_cast<int32_t>(micros / micros_per_hr);
+}
+} // namespace
+
+int32_t hour_transform_visitor::operator()(const primitive_value& v) {
+    if (std::holds_alternative<time_value>(v)) {
+        return micros_to_hr(std::get<time_value>(v).val);
+    }
+    if (std::holds_alternative<timestamp_value>(v)) {
+        return micros_to_hr(std::get<timestamp_value>(v).val);
+    }
+    if (std::holds_alternative<timestamptz_value>(v)) {
+        return micros_to_hr(std::get<timestamptz_value>(v).val);
+    }
+    throw std::invalid_argument(
```

Comment:
How do we intend to use this? If the user can set their own partition spec, we will want to validate it before calling a function that can throw on it.

Reply:
IIUC Iceberg should allow us to write with the hour-only spec even if the user has specified other partition specs for the table. I believe it's up to the writer, and it's the writer's job to add their partition spec to table metadata. I.e. it seems like we should be able to set the spec to something that is RP compatible. Down the line maybe we would be more reactive to changes, or allow users to specify partitioning via topic properties.